<a href="https://colab.research.google.com/github/brayannmb/7DaysOfCode-DataScience/blob/main/7DaysOfCode_Day1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Desafio 7 Days of Code - Alura**

Desafio aceito! 

Costumo acompanhar as iniciativas da Alura nas redes sociais e me interesso por novos desafios, ainda mais quando vem diretamente do mestre Paulo Vasconcellos.

No dia 01, o desafio está em fazer a limpeza dos dados do portal [CEAPS](https://www12.senado.leg.br/transparencia/dados-abertos-transparencia/dados-abertos-ceaps?utm_source=ActiveCampaign&utm_medium=email&utm_content=%237DaysOfCode+-+Ci%C3%AAncia+de+Dados+1%2F7%3A+Data+Cleaning+and+Preparation&utm_campaign=%5BAlura+%237Days+Of+Code%5D%28Java%29+Dia+1%2F7%3A+Consumir+uma+API+de+filmes) (Cota para Exercício da Atividade Parlamentar dos Senadores). Este portal contém as informações sobre as declarações dos senadores no Brasil. 

In [1]:
import pandas as pd

In [2]:
df_2022 = pd.read_csv('https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_2022.csv', encoding='1252', sep=';', header=1)

Para poder utilizar o cabeçalho corretamente, foi necessário incluir o parâmetro `header` e escolher a segunda linha do dataframe original.

In [3]:
df_2022.head()

ANO  MES       SENADOR                                       TIPO_DESPESA  \
0  2022    1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
1  2022    1  ACIR GURGACZ                Divulgação da atividade parlamentar   
2  2022    1  ACIR GURGACZ                Divulgação da atividade parlamentar   
3  2022    1  ACIR GURGACZ                Divulgação da atividade parlamentar   
4  2022    1  ACIR GURGACZ                Divulgação da atividade parlamentar   

             CNPJ_CPF                                         FORNECEDOR  \
0      004.948.028-63                      GILBERTO PISELO DO NASCIMENTO   
1  26.320.603/0001-64                                      INFORMANAHORA   
2  13.659.201/0001-47                    LINHA PURPURA FOTO E VIDEO LTDA   
3  23.652.846/0001-01                  ROBERTO GUTIERREZ DA ROCHA M.E.I.   
4  08.941.827/0001-01  RONDONIA DINÂMICA COM. E SERV. DE INFORMÁTICA ...   

           DOCUMENTO        DATA  \
0             001/22  03/01/2022   
1  000000000000310/A  04/01/2022   
2                107  14/01/2022   
3                187  18/01/2022   
4  000000000001772/A  17/01/2022   

                                        DETALHAMENTO VALOR_REEMBOLSADO  \
0  Despesa com pagamento de aluguel de imóvel par...              6000   
1  Despesa com divulgação da atividade parlamenta...              1500   
2  Despesa com produção de texto e edição de víde...              6000   
3                Divulgação da atividade parlamentar              1000   
4                Divulgação da atividade parlamentar              2000   

   COD_DOCUMENTO  
0        2173614  
1        2173615  
2        2173616  
3        2173618  
4        2173617

In [4]:
df_2022.tail()

ANO  MES           SENADOR  \
7038  2022    5  ZEQUINHA MARINHO   
7039  2022    5  ZEQUINHA MARINHO   
7040  2022    5  ZEQUINHA MARINHO   
7041  2022    5  ZEQUINHA MARINHO   
7042  2022    5  ZEQUINHA MARINHO   

                                           TIPO_DESPESA            CNPJ_CPF  \
7038  Passagens aéreas, aquáticas e terrestres nacio...  22.052.777/0001-32   
7039  Passagens aéreas, aquáticas e terrestres nacio...  22.052.777/0001-32   
7040  Passagens aéreas, aquáticas e terrestres nacio...  22.052.777/0001-32   
7041  Passagens aéreas, aquáticas e terrestres nacio...  22.052.777/0001-32   
7042  Passagens aéreas, aquáticas e terrestres nacio...  22.052.777/0001-32   

         FORNECEDOR DOCUMENTO        DATA  \
7038  Exceller Tour    WRREKJ  17/05/2022   
7039  Exceller Tour    SKP1MQ  17/05/2022   
7040  Exceller Tour    WWDZKA  18/05/2022   
7041  Exceller Tour    FMDGER  25/05/2022   
7042  Exceller Tour    CXFMUH  27/05/2022   

                                           DETALHAMENTO VALOR_REEMBOLSADO  \
7038  Companhia Aérea: GOL, Localizador: WRREKJ. Pas...           2184,54   
7039  Companhia Aérea: AZUL, Localizador: SKP1MQ. Pa...           2903,15   
7040  Companhia Aérea: GOL, Localizador: WWDZKA. Pas...           3104,87   
7041  Companhia Aérea: GOL, Localizador: FMDGER. Pas...           2809,47   
7042  Companhia Aérea: GOL, Localizador: CXFMUH. Pas...            925,62   

      COD_DOCUMENTO  
7038        2180363  
7039        2180365  
7040        2180451  
7041        2180810  
7042        2181308

In [5]:
df_2022.sample(10)

ANO  MES                  SENADOR  \
3947  2022    5           LASIER MARTINS   
5278  2022    5           PLÍNIO VALÉRIO   
6495  2022    1      WELLINGTON FAGUNDES   
229   2022    4       ALEXANDRE SILVEIRA   
2669  2022    3                 GIORDANO   
6485  2022    4  VENEZIANO VITAL DO RÊGO   
288   2022    2           ANGELO CORONEL   
2800  2022    3           HUMBERTO COSTA   
1053  2022    4           CONFÚCIO MOURA   
5049  2022    3               PAULO PAIM   

                                           TIPO_DESPESA            CNPJ_CPF  \
3947  Passagens aéreas, aquáticas e terrestres nacio...  07.575.651/0001-59   
5278  Passagens aéreas, aquáticas e terrestres nacio...  16.978.175/0001-08   
6495  Aquisição de material de consumo para uso no e...  18.804.888/0001-80   
229   Aluguel de imóveis para escritório político, c...  40.432.544/0112-62   
2669  Locomoção, hospedagem, alimentação, combustíve...  33.812.076/0001-70   
6485  Passagens aéreas, aquáticas e terrestres nacio...  07.575.651/0001-59   
288   Aluguel de imóveis para escritório político, c...  35.184.919/0001-67   
2800  Locomoção, hospedagem, alimentação, combustíve...  34.256.587/0001-16   
1053                Divulgação da atividade parlamentar  30.673.595/0001-07   
5049  Locomoção, hospedagem, alimentação, combustíve...  09.146.749/0001-16   

                                             FORNECEDOR        DOCUMENTO  \
3947                                                GOL           LUDZRW   
5278                            ADRIA VIAGENS E TURISMO           YJRBLI   
6495  Brasil One Serviços de Tecnologia da Informaçã...              236   
229                                           CLARO S.A    2205901732793   
2669                            Pizzaria Bella Forneira              NaN   
6485                                                GOL           IXHURH   
288                    JCAR Locadora de Automóveis Ltda  202200000000004   
2800                 Bardana Cozinha e Restaurante Ltda            53323   
1053                             AMAURI DE SOUZA JUNIOR              331   
5049                    Autolocadora Linck e Mello Ltda            22904   

            DATA                                       DETALHAMENTO  \
3947  13/05/2022  Companhia Aérea: GOL, Localizador: LUDZRW. Pas...   
5278  19/05/2022  Companhia Aérea: GOL, Localizador: YJRBLI. Pas...   
6495  03/03/2022  Despesa com serviços de informática, referente...   
229   04/05/2022                                                NaN   
2669  29/03/2022                                                NaN   
6485  13/04/2022  Companhia Aérea: GOL, Localizador: IXHURH. Pas...   
288   01/03/2022  Locacao de carro de apoio a atividade parlamen...   
2800  30/03/2022        Despesa com alimentação para o parlamentar.   
1053  27/04/2022  Despesa com prestação de serviços de marketing...   
5049  06/04/2022  Locação de veículo - carro utilizado no exercí...   

     VALOR_REEMBOLSADO  COD_DOCUMENTO  
3947           2225,67        2181640  
5278           3477,17        2180732  
6495              3820        2175452  
229             421,79        2179929  
2669               219        2179067  
6485           1167,23        2179629  
288               9900        2175448  
2800             69,86        2177257  
1053              5500        2179457  
5049              3000        2178559

Com base nas visualizações acima, é possível identificar algumas inconsistências nos dados, como três colunas com a mesma finalidade de representar a data e diversos valores nulos para a variável de detalhamento.

O próximo passo é verificar como estão os tipos de dados de cada uma das variáveis.

In [6]:
df_2022.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ANO                7043 non-null   int64 
 1   MES                7043 non-null   int64 
 2   SENADOR            7043 non-null   object
 3   TIPO_DESPESA       7043 non-null   object
 4   CNPJ_CPF           7043 non-null   object
 5   FORNECEDOR         7043 non-null   object
 6   DOCUMENTO          6694 non-null   object
 7   DATA               7043 non-null   object
 8   DETALHAMENTO       3737 non-null   object
 9   VALOR_REEMBOLSADO  7043 non-null   object
 10  COD_DOCUMENTO      7043 non-null   int64 
dtypes: int64(3), object(8)
memory usage: 605.4+ KB


Com base nos resultados apresentados do comando `.info()`, as colunas `Detalhamento` e `Documento` possuem valores inválidos. O tipo de dado variável `Data` foi identificada e classificada incorretamente.

Para corrigir os problemas apresentados agora, seguirei algumas etapas, como remoção de colunas, alteração do tipo de dados a análise dos valores nulos.  

In [7]:
df_clean = df_2022.copy()

In [8]:
df_clean.drop(columns=['ANO', 'MES'], inplace=True)

In [9]:
df_clean.columns

Index(['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO',
       'DATA', 'DETALHAMENTO', 'VALOR_REEMBOLSADO', 'COD_DOCUMENTO'],
      dtype='object')

In [10]:
df_clean['DATA'] = pd.to_datetime(df_clean['DATA'], format='%d/%m/%Y')

In [11]:
df_clean['VALOR_REEMBOLSADO'] = df_clean['VALOR_REEMBOLSADO'].apply(lambda x: x.replace(',', '.'))
df_clean['VALOR_REEMBOLSADO'] = df_clean['VALOR_REEMBOLSADO'].astype('float64') 

In [12]:
df_clean[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']] = df_clean[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']].astype('string') 

In [51]:
def combine_dfs(dataframe, year):
  df =  pd.read_csv(f'https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_{year}.csv', encoding='1252', sep=';', header=1)
  df['SENADOR'] = df['SENADOR'].apply(lambda x: x.title())
  df['FORNECEDOR'] = df['FORNECEDOR'].apply(lambda x: x.title())
  df_aux = pd.read_csv(f'https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_{year}.csv', encoding='1252', sep=';', header=1) 
  df_aux = pd.to_datetime(df_aux['DATA'], format='%d/%m/%Y', errors='coerce')
  error_values = df_aux[df_aux['DATA'].isnull()].index.tolist()
  for error in error_values:
    try:
      df.loc[error, 'DATA']  = df.loc[error, 'DATA'].replace('/02', '20')
    except Exception as e:
      print(e)
  df_final = structure_data(df)
  dataframe = dataframe.append(df_final)
  return dataframe
  

In [50]:
def structure_data(df):
  try:
    df['SENADOR'] = df['SENADOR'].apply(lambda x: x.title())
    df['FORNECEDOR'] = df['FORNECEDOR'].apply(lambda x: x.title())
    df['VALOR_REEMBOLSADO'] = df['VALOR_REEMBOLSADO'].apply(lambda x: x.replace(',', '.'))
    df['VALOR_REEMBOLSADO'] = df['VALOR_REEMBOLSADO'].astype('float64')
    df[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']] = df[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']].astype('string')
    df.drop(columns=['ANO', 'MES'], inplace=True)
    return df
  except Exception as e:
    return print(e)

In [48]:
import re
years = ['2019', '2020', '2021', '2022']
for year in years:
  df = pd.read_csv(f'https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_{year}.csv', encoding='1252', sep=';', header=1)
  try:
    pd.to_datetime(df['DATA'], format='%d/%m/%Y')
  except Exception as e:
    print(e)
  for x in df['DATA']:
    if re.search(r'02[0-9]{2}$', x):
      print(x)

Out of bounds nanosecond timestamp: 219-01-31 00:00:00
31/01/0219
Out of bounds nanosecond timestamp: 202-05-05 00:00:00
05/05/0202
Out of bounds nanosecond timestamp: 202-10-29 00:00:00
29/10/0202


In [14]:
df_clean['SENADOR'] = df_clean['SENADOR'].apply(lambda x: x.title())

In [15]:
df_clean['FORNECEDOR'] = df_clean['FORNECEDOR'].apply(lambda x: x.title())

In [16]:
df_2019 = pd.read_csv('https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_2019.csv', encoding='1252', sep=';', header=1)

In [17]:
try:
  df_2019['DATA'] = pd.to_datetime(df_2019['DATA'], format='%d/%m/%Y')
except Exception as e:
  print(e)

Out of bounds nanosecond timestamp: 219-01-31 00:00:00


Quando encontrei o problema cima, confesso que não soube o que fazer de imediato. Diversas tentativas de encontrar o index que correspondia o valor do erro encontrado. 

Uma alternativa, foi utilizar o parâmetro `errors = 'coerce'`, pois este parâmetro transforma os erros em valores nulos. Dessa maneira, através do valor nulo saberia o index do dado inconsistente.

In [18]:
df_2019['DATA'] = pd.to_datetime(df_2019['DATA'], format='%d/%m/%Y', errors='coerce')

In [19]:
error_values = df_2019[df_2019['DATA'].isnull()].index.tolist()

In [20]:
error_values

[10986]

In [21]:
df_2019 = pd.read_csv('https://raw.githubusercontent.com/brayannmb/7DaysOfCode-DataScience/main/despesa_ceaps_2019.csv', encoding='1252', sep=';', header=1)

In [23]:
df_2019.loc[10987, 'DATA']

'03/01/2019'

In [24]:
for error in error_values:
  print(f"Erro encontrado: {df_2019.loc[error, 'DATA']}")

Erro encontrado: 31/01/0219


O valor de data foi inserido incorretamente. `0219`ao invés de `2019`.

In [25]:
df_2019.loc[error_values[0], 'DATA'] = df_2019.loc[error_values[0], 'DATA'].replace('02', '20')

In [28]:
df_2019['DATA'] = pd.to_datetime(df_2019['DATA'], format='%d/%m/%Y')

In [31]:
df_2019.drop(columns=['ANO', 'MES'], axis=1, inplace=True)

In [34]:
df_2019[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']] = df_2019[['SENADOR', 'TIPO_DESPESA', 'CNPJ_CPF', 'FORNECEDOR', 'DOCUMENTO', 'DETALHAMENTO', 'COD_DOCUMENTO']].astype('string') 

In [36]:
try:
  df_2019['VALOR_REEMBOLSADO'].astype('float64')
except Exception as e:
  print(e)

could not convert string to float: '66,02'


In [37]:
df_2019['VALOR_REEMBOLSADO'] = df_2019['VALOR_REEMBOLSADO'].astype('string')

In [38]:
df_2019['VALOR_REEMBOLSADO'] = df_2019['VALOR_REEMBOLSADO'].apply(lambda x: x.replace(',', '.'))

In [40]:
try:
  df_2019['VALOR_REEMBOLSADO'].astype('float64')
  print('Done')
except Exception as e:
  print(e)

Done


In [41]:
df_2019['VALOR_REEMBOLSADO'] = df_2019['VALOR_REEMBOLSADO'].astype('float64')

In [42]:
df_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21634 entries, 0 to 21633
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   SENADOR            21634 non-null  string        
 1   TIPO_DESPESA       21634 non-null  string        
 2   CNPJ_CPF           21634 non-null  string        
 3   FORNECEDOR         21634 non-null  string        
 4   DOCUMENTO          20900 non-null  string        
 5   DATA               21634 non-null  datetime64[ns]
 6   DETALHAMENTO       13092 non-null  string        
 7   VALOR_REEMBOLSADO  21634 non-null  float64       
 8   COD_DOCUMENTO      21634 non-null  string        
dtypes: datetime64[ns](1), float64(1), string(7)
memory usage: 1.5 MB


In [43]:
df_2019.head()

SENADOR                                       TIPO_DESPESA  \
0  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
1  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
2  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
3  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   
4  ACIR GURGACZ  Aluguel de imóveis para escritório político, c...   

             CNPJ_CPF                     FORNECEDOR  DOCUMENTO       DATA  \
0  05.914.650/0001-66                       ENERGISA  006582758 2019-01-04   
1  05.914.650/0001-66                       ENERGISA  006582755 2019-01-04   
2      004.948.028-63  GILBERTO PISELO DO NASCIMENTO      00119 2019-01-07   
3  05.423.963/0001-11                  OI MÓVEL S.A.   86161151 2018-12-25   
4  05.914.650/0001-66                       ENERGISA  007236036 2019-02-04   

                                        DETALHAMENTO  VALOR_REEMBOLSADO  \
0  Despesa com pagamento de energia elétrica do e...              66.02   
1  Despesa com pagamento de energia elétrica do e...             139.98   
2  Despesa com pagamento de aluguel de imóvel par...            6000.00   
3  Despesa com pagamento de telefonia para o escr...             316.39   
4  Despesa com pagamento de energia elétrica para...              99.45   

  COD_DOCUMENTO  
0       2116543  
1       2116546  
2       2113817  
3       2116541  
4       2116550